statescript/store.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2016 Mender Software AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package statescript
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/pkg/errors"
+)
+
+type Store struct {
+	location string
+}
+
+func NewStore(destination string) *Store {
+	return &Store{
+		location: destination,
+	}
+}
+
+func (s *Store) Clear() error {
+	if s.location == "" {
+		return nil
+	}
+	// for safety reasons we are rejecting paths which might harm your system
+	// (/) or paths which are not absolute
+	if s.location == "/" || !filepath.IsAbs(s.location) {
+		return errors.Errorf("Invalid scripts directory path: %s", s.location)
+	}
+
+	err := os.RemoveAll(s.location)
+	if err == nil || (err != nil && os.IsNotExist(err)) {
+		return os.MkdirAll(s.location, 0755)
+	}
+	return err
+}
+
+func (s *Store) StoreScript(r io.Reader, name string) error {
+	sLocation := filepath.Join(s.location, name)
+	f, err := os.OpenFile(sLocation, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0755)
+	if err != nil {
+		return errors.Wrapf(err,
+			"statescript: can not create script file: %v", sLocation)
+	}
+	defer f.Close()
+
+	_, err = io.Copy(f, r)
+	if err != nil {
+		return errors.Wrapf(err,
+			"statescript: can not write script file: %v", sLocation)
+	}
+	return nil
```

Comment:
Btw, it just occurred to me that we need to call `f.Sync()` here, because if we don't, the U-Boot environment may be synced first, and if we lose power at that moment, we will boot into the new update *without* the scripts. So I suggest adding `f.Sync()` before `return nil`.